Update documentation for preview features moving to stable - Actor Reentrancy and State Encryption (#2232)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/actors/actor-reentrancy.md

description: Learn more about actor reentrancy
---

## Actor reentrancy

> A core tenet of the virtual actor pattern is the single-threaded nature of actor execution. Before reentrancy, this caused the Dapr runtime to lock an actor on any given request. A second request could not start until the first had completed. This behavior means an actor cannot call itself, or have another actor call into it even if it is part of the same chain. Reentrancy solves this by allowing requests from the same chain or context to re-enter into an already locked actor. Examples of chains that reentrancy allows can be seen below:

Reviewer: ```suggestion
A core tenet of the virtual actor pattern is the single-threaded nature of actor execution. Without reentrancy, the the Dapr runtime locks on all actor requests, even those that are in the same call chain. A second request could not start until the first had completed. This means an actor cannot call itself, or have another actor call into it even if it is part of the same chain. Reentrancy solves this by allowing requests from the same chain, or context, to re-enter into an already locked actor. This is especially useful in scenarios where an actor wants to call a method on itself or when actors are used in workflows where other actors are used to perform work, and they then call back onto the coordinating actor. Examples of chains that reentrancy allows can be seen below:
```

description: Learn more about actor reentrancy
---

## Actor reentrancy

> A core tenet of the virtual actor pattern is the single-threaded nature of actor execution. Without reentrancy, the the Dapr runtime locks on all actor requests, even those that are in the same call chain. A second request could not start until the first had completed. This means an actor cannot call itself, or have another actor call into it even if it is part of the same chain. Reentrancy solves this by allowing requests from the same chain, or context, to re-enter into an already locked actor. This is especially useful in scenarios where an actor wants to call a method on itself or when actors are used in workflows where other actors are used to perform work, and they then call back onto the coordinating actor. Examples of chains that reentrancy allows can be seen below:

Reviewer: ```suggestion
A core tenet of the virtual actor pattern is the single-threaded nature of actor execution. Without reentrancy, the Dapr runtime locks on all actor requests, even those that are in the same call chain. A second request could not start until the first had completed. This means an actor cannot call itself, or have another actor call into it even if it is part of the same chain. Reentrancy solves this by allowing requests from the same chain, or context, to re-enter into an already locked actor. This is especially useful in scenarios where an actor wants to call a method on itself or when actors are used in workflows where other actors are used to perform work, and they then call back onto the coordinating actor. Examples of chains that reentrancy allows are shown below:
```

Reviewer: @amulyavarote - Looks good. One small typo above. PLease resolve the conflicts.

With reentrancy, there can be more complex actor calls without sacrificing the single-threaded behavior of virtual actors.

> The `maxStackDepth` parameter gets a value that controls how many reentrant calls be made to the same actor.

Reviewer: ```suggestion
The `maxStackDepth` parameter sets a value that controls how many reentrant calls be made to the same actor. By default this is set to 32, which is sufficient in most cases. 
```

## Enable Actor Reentrancy with Actor Configuration

> The actor that will be reentrant must provide the appropriate configuration to use reentrancy. This is done by the actor's endpoint for `GET /dapr/config`, similar to other actor configuration elements. Here is a snipet of an actor written in Golang providing the configuration:

Reviewer: ```suggestion
The actor that will be reentrant must provide configuration to use reentrancy. This is done by the actor's endpoint for `GET /dapr/config`, similar to other actor configuration elements. 
```

> {{< tabs Go Dotnet Python JavaScript >}}

Reviewer: ```suggestion
{{< tabs Dotnet Java Python Go JavaScript >}}
```
Please put the code tabs into this order and also add a blank entry for Java, with a sentence saying "Reentrancy has not yet been included into the Java SDK"

{{% codetab %}}

```go

> type daprConfig struct {

Reviewer: ```suggestion
Here is a snippet of an actor written in Golang providing the reentrancy configuration via the HTTP API. Reentrancy has not yet been included into the Go SDK.

type daprConfig struct {
```

With reentrancy, there can be more complex actor calls without sacrificing the single-threaded behavior of virtual actors.

## Enabling actor reentrancy
Actor reentrancy is a one step process.

> ### Actor runtime configuration

Reviewer 2: Instead of having a header with just the one line in it, I'd have just one header like `Enable Actor Reentrancy with Actor Configuration`.
